Add Ethereum Smart Contract binary bytecode matching

Alexvozhak opened the pull request:
**Description**:

This PR enhances `diffyscan` by integrating Ethereum smart contract binary matching. It incorporates functionalities from [verify-bytecode.ts](https://github.com/lidofinance/stonks/blob/main/scripts/verify-bytecode.ts) and [bytecode_verificator.sh](https://github.com/lidofinance/lido-dao/blob/master/bytecode-verificator/bytecode_verificator.sh).

---

**Key Changes**:

- **Showcases**:
  - **Sepolia Network**: Added support for all contracts by extending [`lido_dao_sepolia_config.json`](https://github.com/lidofinance/diffyscan/blob/main/config_samples/lido_dao_sepolia_config.json).
  - **Holesky Network**: Added support for 3 out of 53 contracts by extending [`lido_dao_holesky_config.json`](https://github.com/lidofinance/diffyscan/blob/main/config_samples/lido_dao_holesky_config.json).

- **New Modules and Refactoring**:
  - **Binary Comparison Logic**: Introduced `binary_verifier.py` to handle binary comparison of contracts.
  - **Argument Encoding**: Added `encoder.py` for encoding contract arguments, referencing the [Solidity ABI specification](https://docs.soliditylang.org/en/develop/abi-spec.html).
  - **Compiler Module**: Moved all source compilation code to `compiler.py`.
  - **Calldata Handling**: Created `calldata.py` for calculating and parsing constructor arguments from configurations.
  - **Type Definitions**: Renamed and extended `types.py` to `custom_types.py`.
  - **Node Interactions**:
    - Encapsulated interactions with the local Hardhat node in `hardhat.py`.
    - Moved common node operations to `node_handler.py`.
  - **Exception Handling**: Added specific exceptions for new features in `custom_exceptions.py`.

- **Configurations and Documentation**:
  - **Configuration Updates**:
    - Extended `lido_dao_sepolia_config.json` and `lido_dao_holesky_config.json` with a `binary_comparison` section.
    - Added specific Hardhat configurations for Sepolia and Holesky networks.
  - **Documentation**:
    - Updated the README to reflect new features and usage instructions.
    - Included support for user-defined Hardhat configurations as shown in the README example.

- **Improvements and Fixes**:
  - **Code Quality**:
    - Fixed existing errors identified by the `mypy` linter.
    - Enhanced code readability and organization.
  - **Backwards compatibility**: Ensured that existing `diffyscan` usage remains unaffected.

---

**Implemented Logic for Contract Code Matching**:

1. **Source Retrieval**: Obtain contract sources from Etherscan.
2. **Source Diffing**: Perform source code diffing as previously implemented.
3. **Bytecode Matching**: Compile the contracts and attempt to match the bytecode obtained from the RPC node.
4. **Matching Outcome**:
   - **Success**: If the bytecode matches, proceed to the next contract.
   - **Metadata Differences**: If only the metadata differs, display a warning but continue processing.
5. **Error Handling**:
   - **No Immutables**: If there are no immutables and the bytecode doesn't match, display an error.
   - **With Immutables**: If immutables are present and the configuration includes constructor calldata or parameters, attempt to deploy the contract and compare the resulting bytecode.

---

By integrating these enhancements, `diffyscan` has been enriched with the ability to verify smart contracts at the bytecode level, ensuring greater reliability and security in contract deployment and verification processes.

tamtamchik commented on README.md:
Locking on ganache (outdated and even archived https://github.com/trufflesuite/ganache) looks like a major issue. I think a node-agnostic setup with an anvil/hardhat node running on a specific port will be better.

tamtamchik replied:
I have a request: can you please format the code in `Black` style so it's easier to read?

tamtamchik commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,333 @@
+def prepare_compiler(platform, build_info, compiler_path):
+    create_dirs(compiler_path)
+    compiler_binary = download_compiler(platform, build_info, compiler_path)
+    valid_checksum = build_info['sha256'][2:];
```
```suggestion
    valid_checksum = build_info['sha256'][2:]
```

Alexvozhak replied:
done

tamtamchik commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,333 @@
+  return contract_address
+    
+          
+OPCODES = {
```
Move to constants?

Alexvozhak replied:
done

tamtamchik commented on diffyscan/utils/common.py:
```diff
@@ -43,7 +40,21 @@ def fetch(url, headers={}):
+    return response
+
+
+def fetch(url, headers={}):
```
```suggestion
def fetch(url, headers=None):
```

Alexvozhak replied:
done

tamtamchik commented on diffyscan/utils/common.py:
```diff
@@ -43,7 +40,21 @@ def fetch(url, headers={}):
+    return handle_response(response, url)
+
+
+def pull(url, payload={}):
```
```suggestion
def pull(url, payload=None):
```

Alexvozhak replied:
done

tamtamchik commented on diffyscan/utils/common.py:
```diff
@@ -57,3 +68,13 @@ def parse_repo_link(repo_link):
+    if platform_name == "linux":
+        return "linux-amd64"
+    elif platform_name == "darwin":
+        return "macosx-amd64"
```
It can also be `macosx-arm64` (for Apple M chip based)

tamtamchik commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,333 @@
+
+    checkpointsArray = sorted(list(checkpoints))
+
+    def hex(index, padStart=2):
```
Move to logger?

Alexvozhak replied:
done

tamtamchik reviewed (changes requested):
Round 2: Added a bunch of new comments and suggestions. 

Run the testing command: 
```
diffyscan config_samples/lido_dao_sepolia_config.json hardhat_configs/sepolia_hardhat.config.js --binary-check -y
```
And I see some errors in logs:

```
🔵 [INFO] Receiving the account from "http://127.0.0.1:7545" ...
 🟢 [OKAY] The account was successfully received
 🔵 [INFO] Started binary checking for 0x8A1AA86d35b2EE8C9369618E7D7b40000cCD3295
 🟢 [OKAY] Contracts were successfully compiled. The target contract is ready for matching
 🔵 [INFO] Constructor args types: ['address', 'bytes32', 'bytes']
 🔵 [INFO] Trying to deploy transaction to "http://127.0.0.1:7545" ...
 🔴 [ERROR] Failed to deploy transaction: Error: Transaction reverted without a reason string
 🔴 [ERROR] Failed to deploy bytecode to http://127.0.0.1:7545, binary diff skipped (0x8A1AA86d35b2EE8C9369618E7D7b40000cCD3295)```
 
I think such errors should not be skipped, and checking should be stopped on such errors.

Alexvozhak commented on diffyscan/diffyscan.py:
```diff
@@ -7,35 +7,104 @@
 g_skip_user_input: bool = False
 
+
 def prettify_solidity(solidity_contract_content: str):
```
done

Alexvozhak commented on diffyscan/utils/constants.py:
```diff
@@ -1,8 +1,25 @@
 DEFAULT_CONFIG_PATH = "config.json"
+
+REMOTE_RPC_URL = os.getenv("REMOTE_RPC_URL", "")
+if not REMOTE_RPC_URL:
```
done

Alexvozhak commented on diffyscan/utils/constants.py:
```diff
@@ -1,8 +1,25 @@
+    raise ValueError("REMOTE_RPC_URL variable is not set")
+
+SOLC_DIR = os.getenv("SOLC_DIR", "")
+if SOLC_DIR == "":
```
done

Alexvozhak commented on diffyscan/utils/constants.py:
```diff
@@ -1,8 +1,25 @@
+if SOLC_DIR == "":
+    raise ValueError("SOLC_DIR variable is not set")
+
+LOCAL_RPC_URL = os.getenv("LOCAL_RPC_URL", "")
```
done

Alexvozhak commented on README.md:
````diff
@@ -34,11 +35,22 @@ Set your Github token to query API without strict rate limiting,
+Set absolute or relative path for downloading solc compilers
+
+```bash
+export SOLC_DIR=<solc-path>
````
done

Alexvozhak commented on README.md:
````diff
@@ -5,17 +5,18 @@
 ```
 
-If need `--prettify` option
+If need `--prettify`, '--binary-check' options
````
done

Alexvozhak commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,333 @@
+    compilers_list_url = (
+      f'https://raw.githubusercontent.com/ethereum/solc-bin/gh-pages/{platform}/list.json'   
+    )
+    available_compilers_list = fetch(compilers_list_url).json()
```
done

Alexvozhak commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,333 @@
+    )
+    download_compiler_response = fetch(compiler_url)  
+
+    with open(destination_path, 'wb') as compiler_file:
```
done

Alexvozhak commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,333 @@
+    os.chmod(compiler_path, compiler_file_rights.st_mode | stat.S_IEXEC)
+    
+def compile_contracts(compiler_path, input_settings):
+    process = subprocess.run([compiler_path,'--standard-json'], input=input_settings.encode(), capture_output=True, check=True, timeout=30)
```
done

Alexvozhak commented on diffyscan/diffyscan.py:
```diff
@@ -145,32 +213,94 @@ def run_diff(config, name, address, explorer_api_token, github_api_token, recurs
+    parser.add_argument(
+        "--binary-check",
+        "-b",
+        help="Match contracts by binaries such as verify-bytecode.ts",
```
done

Alexvozhak commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,333 @@
+def check_compiler_checksum(compiler, valid_checksum):
+    compiler_checksum = hashlib.sha256(compiler).hexdigest()
+    if compiler_checksum != valid_checksum:
+      raise ValueError(f'Bad checksum')
```
done

Alexvozhak commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,333 @@
+    required_build_info = next((compiler for compiler in available_compilers_list['builds'] if compiler['longVersion'] == required_compiler_version), None)
+
+    if not required_build_info:
+      raise ValueError(f'Required compiler version for "{platform}" is not found')
```
For example?

tamtamchik replied:
For this example, it may look something like this. This is a "nice to have" feature since `diffyscan,` I assume, can be used as a module and therefore should produce meaningful error types, not generic ones.  Disclaimer: I might be wrong, and this is not in the scope. cc @TheDZhon 

```python
class CompilerVersionNotFoundError(Exception):
    def __init__(self, platform, version):
        self.platform = platform
        self.version = version
        self.message = f'Required compiler version "{version}" for platform "{platform}" is not found'
        super().__init__(self.message)
        
...
raise CompilerVersionNotFoundError(platform, required_compiler_version)
```

Alexvozhak commented on diffyscan/utils/binary_verifier.py:
done

Alexvozhak commented on README.md:
````diff
@@ -34,11 +35,22 @@ Set your Github token to query API without strict rate limiting,
 
 ```bash
-diffyscan config_samples/lido_dao_sepolia_config.json
+diffyscan config_samples/lido_dao_sepolia_config.json hardhat_configs/sepolia_hardhat.config.js
````
I'm not sure that I correctly understood how it should work

tamtamchik replied:
Basically, you can have one `hardhat.config.js` file that will be automatically used by hardhat and setup all the networks there (example, https://github.com/lidofinance/core/blob/master/hardhat.config.ts#L40-L68), then just specify `--network sepolia` on node start and HH will automatically use config from this file.

Alexvozhak replied:
It won't work like that, [proof](https://github.com/NomicFoundation/hardhat/issues/2449#issuecomment-1059259560)

tamtamchik replied:
@Alexvozhak roger that! 🫡

Alexvozhak commented on diffyscan/utils/hardhat.py:
```diff
@@ -0,0 +1,77 @@
+            f"npx hardhat node --hostname {parsed_url.hostname} "
+            f"--port {parsed_url.port} "
+            f"--config {hardhat_config_relative_path} "
+            f"--fork {binary_config["remote_RPC_URL"]}"
```
@tamtamchik after adding sepolia fork deploy stopped to work, this problem hasn't solved yet

tamtamchik replied:
@Alexvozhak thank you for the update

Alexvozhak commented:
@tamtamchik hardhat needs to be delivered manually (npm install hardhat), i will try to fix it

Alexvozhak commented:
@tamtamchik hardhat autoinstall has been fixed

tamtamchik commented:
JFYI, I have changed https://ethereum-sepolia-rpc.publicnode.com to https://sepolia.drpc.org, and now it runs OK at least for Sepolia.

Alexvozhak commented:
@tamtamchik i have already updated PR with new URL, it really works!

> JFYI, I have changed https://ethereum-sepolia-rpc.publicnode.com to https://sepolia.drpc.org, and now it runs OK at least for Sepolia, no errors.
> 
> I plan to review it again tomorrow and probably approve it if no showstoppers are found. @Alexvozhak, great work! Thank you!

tamtamchik commented on README.md:
```diff
@@ -5,17 +5,18 @@
 [![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
 
-Diff your Ethereum smart contracts code from GitHub against Blockchain explorer verified source code.
+Diff your Ethereum smart contracts code from GitHub against Blockchain explorer verified source code, deployed bytecode from local node (Hardhat) against remote
```
Looks unfinished.

Alexvozhak replied:
done

tamtamchik commented on README.md:
```diff
@@ -5,17 +5,18 @@
+Diff your Ethereum smart contracts code from GitHub against Blockchain explorer verified source code, deployed bytecode from local node (Hardhat) against remote
 
 Supports reformatting solidity code by means of prettifier solidity plugin before comparing the sources (option `--prettify`).
+Supports binary checking deployed bytecode from local node (Hardhat) against remote (see section 'binary_checking' in lido_dao_sepolia_config.json as example)
```
```suggestion
Supports binary checking deployed bytecode from the local node (Hardhat) against remote (see section 'binary_checking' in `./config_samples/lido_dao_sepolia_config.json` as an example)
```

tamtamchik commented on README.md:
````diff
@@ -5,17 +5,18 @@
 ```
 
-If need `--prettify` option
+If need `--prettify` or binary checking options support 
````
```suggestion
If need `--prettify` or binary checking options support:
```

tamtamchik commented on README.md:
```diff
@@ -90,7 +91,7 @@ This project was developed using these dependencies with their exact versions li
 - Python 3.12
 - Poetry 1.8
-- if need `--prettify` option support:
+- if need `--prettify` or binary checking options support
```
```suggestion
- if need `--prettify` or binary checking options support:
```

tamtamchik commented on README.md:
````diff
@@ -128,7 +129,7 @@ poetry shell
 ```
 
-5. If need `--prettify` option
+5. If need `--prettify` or binary checking options support
````
```suggestion
5. If need `--prettify` or binary checking options support:
```

tamtamchik commented on config_samples/lido_dao_holesky_config.json:
```diff
@@ -111,5 +112,38 @@
+    },
+    "binary_checking": {
+	"use": 1,
+	"skip_deploy_error" : 1,
```
It should be false by default, I think.

Alexvozhak replied:
done

tamtamchik commented on config_samples/lido_dao_holesky_config.json:
```diff
@@ -111,5 +112,38 @@
         }
+    },
+    "binary_checking": {
+	"use": 1,
```
```suggestion
	"enable": true,
```

tamtamchik replied:
Should be handled in the code accordingly.

Alexvozhak replied:
done

tamtamchik commented on config_samples/lido_dao_holesky_config.json:
```diff
@@ -111,5 +112,38 @@
+    },
+    "binary_checking": {
+	"use": 1,
+	"skip_deploy_error" : 1,
```
```suggestion
	"skip_deploy_error" : false,
```

Alexvozhak replied:
done

tamtamchik commented on diffyscan/diffyscan.py:
```diff
@@ -149,28 +217,90 @@ def process_config(path: str, recursive_parsing: bool, unify_formatting: bool):
     if "explorer_token_env_var" in config:
-        explorer_token = load_env(config["explorer_token_env_var"], masked=True, required=False)
+        explorer_token = load_env(
+            config["explorer_token_env_var"], masked=True, required=False
```
All new config values should be properly typed and added to `utils→types→Config`

Alexvozhak replied:
I don't understand what role it plays. If I comment it out or if I add incorrect types, nothing breaks. 
So I suggest to remove it completely

Alexvozhak replied:
done

tamtamchik commented on diffyscan/utils/compiler.py:
```diff
@@ -0,0 +1,109 @@
+        raise ValueError(f"Unsupported platform {platform_name}")
+
+
+def get_compiler_info(platform, required_compiler_version):
```
platform shadows the import platform 
```suggestion
def get_compiler_info(required_platform, required_compiler_version):
```

Alexvozhak replied:
done

tamtamchik commented on diffyscan/utils/compiler.py:
```diff
@@ -0,0 +1,109 @@
+
+
+def get_compiler_info(platform, required_compiler_version):
+    compilers_list_url = f"https://raw.githubusercontent.com/ethereum/solc-bin/gh-pages/{platform}/list.json"
```
```suggestion
    compilers_list_url = f"https://raw.githubusercontent.com/ethereum/solc-bin/gh-pages/{required_platform}/list.json"
```

Alexvozhak replied:
done

tamtamchik commented on diffyscan/utils/compiler.py:
```diff
@@ -0,0 +1,109 @@
+    )
+
+    if not required_build_info:
+        raise ValueError(f'Required compiler version for "{platform}" is not found')
```
```suggestion
        raise ValueError(f'Required compiler version for "{required_platform}" is not found')
```

Alexvozhak replied:
done

tamtamchik commented on diffyscan/utils/compiler.py:
```diff
@@ -0,0 +1,109 @@
+    return required_build_info
+
+
+def download_compiler(platform, build_info, destination_path):
```
```suggestion
def download_compiler(required_platform, build_info, destination_path):
```

Alexvozhak replied:
done

tamtamchik commented on diffyscan/utils/compiler.py:
```diff
@@ -0,0 +1,109 @@
+
+
+def download_compiler(platform, build_info, destination_path):
+    compiler_url = f'https://binaries.soliditylang.org/{platform}/{build_info["path"]}'
```
```suggestion
    compiler_url = f'https://binaries.soliditylang.org/{required_platform}/{build_info["path"]}'
```

tamtamchik commented on diffyscan/utils/compiler.py:
```diff
@@ -0,0 +1,109 @@
+    os.chmod(compiler_path, compiler_file_rights.st_mode | stat.S_IEXEC)
+
+
+def prepare_compiler(platform, build_info, compiler_path):
```
```suggestion
def prepare_compiler(required_platform, build_info, compiler_path):
```

tamtamchik commented on diffyscan/utils/compiler.py:
```diff
@@ -0,0 +1,109 @@
+
+def prepare_compiler(platform, build_info, compiler_path):
+    create_dirs(compiler_path)
+    compiler_binary = download_compiler(platform, build_info, compiler_path)
```
```suggestion
    compiler_binary = download_compiler(required_platform, build_info, compiler_path)
```

Alexvozhak replied:
done

tamtamchik commented on diffyscan/utils/encoder.py:
```diff
@@ -0,0 +1,109 @@
+            encoded_offsets += offset
+        else:
+            logger.warn(
+                f"Unknown constructor argument type '{arg_type}', use --constructor-calldata instead"
```
I can't find any `--constructor-calldata` except this place, how should this work?

Alexvozhak replied:
done. Added field 'constructor_calldata' into config for using prepared calldata, see updates into sepolia and holesky configs

tamtamchik commented on diffyscan/utils/explorer.py:
```diff
@@ -55,16 +79,16 @@ def _get_contract_from_zksync(zksync_explorer_hostname, contract):
     etherscan_link = f"https://{mantle_explorer_hostname}/api?module=contract&action=getsourcecode&address={contract}"
-
-    response = fetch(etherscan_link)
+    response = fetch(etherscan_link).json
```
```suggestion
    response = fetch(etherscan_link).json()
```

tamtamchik commented on diffyscan/utils/explorer.py:
```diff
@@ -15,36 +17,58 @@ def _get_contract_from_etherscan(token, etherscan_hostname, contract):
     )
 
-    response = fetch(zksync_explorer_link)
+    response = fetch(zksync_explorer_link).json
```
```suggestion
    response = fetch(zksync_explorer_link).json()
```

tamtamchik commented on diffyscan/utils/explorer.py:
```diff
@@ -15,36 +17,58 @@ def _get_contract_from_etherscan(token, etherscan_hostname, contract):
+    solc_input = result["SourceCode"]
+
+    if solc_input.startswith("{{"):
+        return {
```
Can we add a returning type to all such functions so it'll be easier to see that the return value is valid?

tamtamchik commented on diffyscan/utils/hardhat.py:
```diff
@@ -0,0 +1,77 @@
+            os.kill(self.sub_process.pid, signal.SIGTERM)
+            logger.info(f"Hardhat stopped, PID {self.sub_process.pid}")
+
+    def get_config_path(from_path: str, to_path: str, filename: str) -> str:
```
`@staticmethod`?

Alexvozhak replied:
done

tamtamchik reviewed (commented):
All in all looks good, but I got an error again: 

`diffyscan ./config_samples/lido_dao_holesky_config.json`

```
 🔵 [INFO] Started binary checking for (0x3F1c547b21f65e10480dE3ad8E19fAAC46C95034 : AppProxyUpgradeable)
 🟢 [OKAY] Contracts were successfully compiled. The target contract is ready for matching
 🔵 [INFO] Hardhat stopped, PID 20053
Traceback (most recent call last):
  File ".../diffyscan-9kJxk0py-py3.12/bin/diffyscan", line 6, in <module>
    sys.exit(main())
             ^^^^^^
  File ".../diffyscan-fork/diffyscan/diffyscan.py", line 316, in main
    process_config(args.path, args.support_brownie, args.prettify)
  File ".../diffyscan-fork/diffyscan/diffyscan.py", line 261, in process_config
    run_binary_diff(
  File ".../diffyscan-fork/diffyscan/diffyscan.py", line 39, in run_binary_diff
    contract_creation_code, immutables = get_code_from_explorer(
                                         ^^^^^^^^^^^^^^^^^^^^^^^
  File ".../diffyscan-fork/diffyscan/utils/explorer.py", line 186, in get_code_from_explorer

ValueError: Failed to find constructor args values 0x3F1c547b21f65e10480dE3ad8E19fAAC46C95034
```

Alexvozhak commented:
> All in all looks good, but I got an error again:
> 
> `diffyscan ./config_samples/lido_dao_holesky_config.json`
> 
> ```
>  🔵 [INFO] Started binary checking for (0x3F1c547b21f65e10480dE3ad8E19fAAC46C95034 : AppProxyUpgradeable)
>  🟢 [OKAY] Contracts were successfully compiled. The target contract is ready for matching
>  🔵 [INFO] Hardhat stopped, PID 20053
> Traceback (most recent call last):
>   File ".../diffyscan-9kJxk0py-py3.12/bin/diffyscan", line 6, in <module>
>     sys.exit(main())
>              ^^^^^^
>   File ".../diffyscan-fork/diffyscan/diffyscan.py", line 316, in main
>     process_config(args.path, args.support_brownie, args.prettify)
>   File ".../diffyscan-fork/diffyscan/diffyscan.py", line 261, in process_config
>     run_binary_diff(
>   File ".../diffyscan-fork/diffyscan/diffyscan.py", line 39, in run_binary_diff
>     contract_creation_code, immutables = get_code_from_explorer(
>                                          ^^^^^^^^^^^^^^^^^^^^^^^
>   File ".../diffyscan-fork/diffyscan/utils/explorer.py", line 186, in get_code_from_explorer
> 
> ValueError: Failed to find constructor args values 0x3F1c547b21f65e10480dE3ad8E19fAAC46C95034
> ```

Error, because now only 3 contracts from 53 are filled in holesky config (constructor's values)

tamtamchik reviewed (approved):
Don't have significant questions or notices. Looks good to me. Approved.

TheDZhon commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,142 @@
+    i = 0
+    while i < len(buffer):
+        opcode = buffer[i]
+        length = 1 + (opcode - 0x5F if 0x5F <= opcode <= 0x7F else 0)
```
why `0x5F` and `0x7F` aren't constants?

tamtamchik replied:
Plus, it is worth adding a check that the actual length is enough for PUSH(X), wdyt?

TheDZhon commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,142 @@
+                "0x" + expected["bytecode"][2:] if len(expected["bytecode"]) > 2 else ""
+            )
+
+            paramsLength = len(expected["bytecode"]) // 2 - 1
```
It seems here we deal with a hex string `0xAABBCCDD` and try to count the bytes inside.
shall it be done via a little helper func to be more reader-friendly?

Alexvozhak replied:
done

TheDZhon commented on diffyscan/utils/calldata.py:
```diff
@@ -0,0 +1,45 @@
+def get_calldata(contract_address_from_config, target_compiled_contract, binary_config):
+    calldata = get_prepared_calldata_from_config(contract_address_from_config, binary_config)
+    if calldata is not None:
+          return calldata, ''
```
```suggestion
          return calldata, None
```
Here, we have a function that returns `calldata` and error (optionally). 

Why return empty text instead of `None` in such a case?

TheDZhon commented on diffyscan/utils/calldata.py:
```diff
@@ -0,0 +1,45 @@
+    if calldata is not None:
+          return calldata, ''
+    
+    calldata, text_error = parse_calldata_from_config(
+        contract_address_from_config,
+        binary_config["constructor_args"],
+        target_compiled_contract,
+    )
+    return calldata, text_error
```
Thinking further, I am a bit puzzled why we return `text_error` and `calldata` together.
It seems that `text_error` might arise only from the parsed config stage. 

Should it be raised as an exception instead?

TheDZhon commented on diffyscan/utils/calldata.py:
```diff
@@ -0,0 +1,45 @@
+def get_constructor_abi(target_compiled_contract):
+    constructor_abi = None
+    try:
+        constructor_abi = [entry["inputs"] for entry in target_compiled_contract['abi'] if entry["type"] == "constructor"][0]
```
looks like this file hasn't been formatted with black, the line is massively long

Alexvozhak replied:
done

TheDZhon commented on diffyscan/utils/calldata.py:
```diff
@@ -0,0 +1,45 @@
+    try:
+        constructor_abi = [entry["inputs"] for entry in target_compiled_contract['abi'] if entry["type"] == "constructor"][0]
+    except IndexError:
+        logger.okay(f'Constructor in ABI not found, calldata calculation skipped')
```
```suggestion
        logger.okay(f'Contract's ABI doesn't have a constructor, calldata calculation skipped')
```

TheDZhon commented on diffyscan/utils/calldata.py:
```diff
@@ -0,0 +1,45 @@
+        logger.okay(f'Constructor in ABI not found, calldata calculation skipped')
+        return None
+
+    logger.okay(f'Constructor in ABI successfully found: {[arg['type'] for arg in constructor_abi]}')
```
might be too verbose to log every step

Alexvozhak replied:
reduced to: "Constructor in ABI successfully found"

TheDZhon commented on diffyscan/utils/calldata.py:
```diff
@@ -0,0 +1,45 @@
+    logger.okay(f'Constructor in ABI successfully found: {[arg['type'] for arg in constructor_abi]}')
+    return constructor_abi
+
+def get_prepared_calldata_from_config(contract_address_from_config, binary_config):
```
```suggestion
def get_raw_calldata_from_config(contract_address_from_config, binary_config):
```

TheDZhon commented on diffyscan/utils/calldata.py:
```diff
@@ -0,0 +1,45 @@
+    prepared_calldata_from_config = calldata_field[contract_address_from_config]
+    return prepared_calldata_from_config
+    
+def parse_calldata_from_config(contract_address_from_config, constructor_args, target_compiled_contract):    
```
If you return composed types, consider providing type hints for the function declarations, e.g.:

```python
def parse_calldata_from_config(
    contract_address_from_config, 
    constructor_args, 
    target_compiled_contract
) → (Optional[str], Optional[str]):
```

TheDZhon replied:
Nevertheless, I am not sure that we should use Optional types and not raising exceptions

The file feels like a mix of both approaches, but I may miss an intended structure

Alexvozhak replied:
reworked with  exceptions usage

TheDZhon commented on diffyscan/utils/common.py:
```diff
@@ -31,19 +33,38 @@ def load_config(path: str) -> Config:
+    except requests.exceptions.Timeout as timeout_err:
+        raise ValueError(f"Timeout error occurred: {timeout_err}")
+    except requests.exceptions.RequestException as req_err:
+        raise ValueError(f"An error occurred: {req_err}")
```
```suggestion
        raise ValueError(f"Request exception occurred: {req_err}")
```

TheDZhon commented on diffyscan/utils/common.py:
```diff
@@ -31,19 +33,38 @@ def load_config(path: str) -> Config:
+    except requests.exceptions.Timeout as timeout_err:
+        raise ValueError(f"Timeout error occurred: {timeout_err}")
+    except requests.exceptions.RequestException as req_err:
+        raise ValueError(f"An error occurred: {req_err}")
```
```suggestion
        raise ValueError(f"Request exception occurred: {req_err}")
```

TheDZhon commented on diffyscan/utils/common.py:
```diff
@@ -57,3 +78,26 @@ def parse_repo_link(repo_link):
+    )
+    with open(github_file_name, "w") as fp:
+        fp.write(solidity_contract_content)
+    prettier_return_code = subprocess.call(
```
```suggestion

    prettier_return_code = subprocess.call(
```

TheDZhon replied:
to make it a bit easier for reading

TheDZhon commented on diffyscan/utils/compiler.py:
```diff
@@ -0,0 +1,111 @@
+    platform_name = sys.platform
+    if platform_name == "linux":
+        return "linux-amd64"
+    elif platform_name == "darwin":
+        return "macosx-amd64" if platform.machine() == "x86_64" else "macosx-arm64"
```
```suggestion
    elif platform_name == "darwin":
        return "macosx-amd64" if platform.machine() == "x86_64" else "macosx-arm64"
    elif plaform_name == "win32":
        return "windows-amd64"
```

TheDZhon commented on diffyscan/utils/compiler.py:
```diff
@@ -0,0 +1,111 @@
+
+    if not required_build_info:
+        raise ValueError(
+            f'Required compiler version for "{required_platform}" is not found'
```
```suggestion
            f'Required compiler version "{required_compiler_version}" for "{required_platform}" is not found'
```

TheDZhon commented on diffyscan/utils/compiler.py:
```diff
@@ -0,0 +1,111 @@
+            timeout=30,
+        )
+    except subprocess.CalledProcessError as e:
+        raise ValueError(f"Error during subprocess execution: {e}")
```
```suggestion
        raise ValueError(f"Error during compiler subprocess execution: {e}")
```

TheDZhon commented on diffyscan/utils/compiler.py:
```diff
@@ -0,0 +1,111 @@
+    except subprocess.CalledProcessError as e:
+        raise ValueError(f"Error during subprocess execution: {e}")
+    except subprocess.TimeoutExpired as e:
+        raise ValueError(f"Process timed out: {e}")
```
```suggestion
        raise ValueError(f"Compiler process timed out: {e}")
```

TheDZhon commented on diffyscan/utils/constants.py:
```diff
@@ -1,8 +1,52 @@
+
+SOLC_DIR = os.path.join(tempfile.gettempdir(), "solc_builds")
+
+GITHUB_API_TOKEN = os.getenv("GITHUB_API_TOKEN", "")
+if not GITHUB_API_TOKEN:
+    raise ValueError("GITHUB_API_TOKEN variable is not set")
```
Sorry, but it doesn't look like a compile-time constant
I would reconsider placing this env var somewhere else

Alexvozhak replied:
done

TheDZhon commented on diffyscan/utils/encoder.py:
```diff
@@ -0,0 +1,105 @@
+    data_length = len(bytes_str) // 2
+    encoded_length = 0
+    if data_length > 0:
+        encoded_length = ((len(bytes_str) - 1) // 64 + 1) * 64
```
magic

Alexvozhak replied:
source of magic [here](https://github.com/lidofinance/lido-dao/blob/a9a9ea50b5be7dba33e06d70cc068252ffd01b52/bytecode-verificator/bytecode_verificator.sh#L239)

Alexvozhak replied:
now it seems more readable

TheDZhon replied:
Can it be commented with a brief explanation of the essence of this operation instead of saying 'taken from there and untested'? :pray:

TheDZhon replied:
```python
# Calculate the length of the hex-encoded 32-bytes padded data 
# since EVM uses 32-byte (256-bit) words
```

TheDZhon commented on diffyscan/utils/encoder.py:
I believe it's better to provide some documentation links here to understand the process of encoding

TheDZhon replied:
This file is rock-hard for those who aren't familiar with the ABI encoder algorithms

Alexvozhak replied:
added link into comments, also i am going to add it into the readme

TheDZhon commented on diffyscan/utils/encoder.py:
```diff
@@ -0,0 +1,105 @@
+                        (len(constructor_calldata) // 64 + 1) * 32, "064x"
+                    )
+                    constructor_calldata += dynamic_type_length
+                    logger.info(f"dynamic_type_length {dynamic_type_length}")
```
do we need this log?

Alexvozhak replied:
removed

TheDZhon commented on diffyscan/utils/encoder.py:
```diff
@@ -0,0 +1,105 @@
+                ]
+                if len(args_tuple_types) and args_tuple_types[0] == "address[]":
+                    dynamic_type_length = format(
+                        (len(constructor_calldata) // 64 + 1) * 32, "064x"
```
magic

Alexvozhak replied:
tried to do more readable

TheDZhon commented on diffyscan/utils/explorer.py:
```diff
@@ -74,32 +93,85 @@ def _get_contract_from_mantle(mantle_explorer_hostname, contract):
+    contract_name_from_etherscan = result["name"]
+    if contract_name_from_etherscan != contract_name_from_config:
+        raise ValueError(
+            f"Contract name in config does not match with Blockchain explorer {contract_address}: {contract_name_from_config} != {contract_name_from_etherscan}",
```
something bad with formatting here :thinking:

Alexvozhak replied:
done

TheDZhon commented on diffyscan/utils/hardhat.py:
```diff
@@ -0,0 +1,78 @@
+            return s.connect_ex((parsed_url.hostname, parsed_url.port)) == 0
+
+    def start(self, main_config_relative_path: str, binary_config):
+        parsed_url = urlparse(binary_config["local_RPC_URL"])
```
usually it's quite a bad idea to store RPC URL inside the config file because it might contain restricted access tokens that shouldn't be published in the sourcecode repo

the approach should be to load those:
- Etherscan token
- RPC URL(s)
- GitHub token

from private non-disclosed env :warning:

Alexvozhak replied:
done

TheDZhon commented on diffyscan/utils/hardhat.py:
```diff
@@ -0,0 +1,78 @@
+        logger.info(f'Trying to start Hardhat: "{local_node_command}"')
+        is_port_used = self.is_port_in_use(parsed_url)
+        if is_port_used:
+            answer = input(f'Port {parsed_url.port} is busy. Fix it? write "yes": ')
```
```suggestion
            answer = input(f'Port {parsed_url.port} is busy. Kill the app instance occupying the port? write "yes": ')
```

TheDZhon commented on README.md:
The readme file should suggest the default config that supports binary bytecode comparison :pray:

Alexvozhak replied:
done!

TheDZhon reviewed (changes requested):
Thank you for taking the lead on implementing this feature #13 ! :+1:

The PR looks solid, and I appreciate the minimal use of external dependencies, which is crucial for avoiding supply chain vulnerabilities. :detective:

I’ve left a few suggestions and concerns in the comments that I believe would strengthen the implementation. Key points to consider:

* **Execution Flow**: We may need to think more deeply about the flow of execution, especially regarding the comparison process that involves checks without deployment and handling immutable logic, see my comment on this:

![image](https://github.com/user-attachments/assets/067f75b2-6e65-4bd4-8918-c5164cc1c2af)


* **Edge Cases**: It would be helpful to account for some edge cases in the diffing process.
* **Documentation**: Please include the necessary documentation references for the ABI encoder.
* **Code Quality**: There’s some room for code improvement, which I’ve highlighted in the comments. :heart:

tamtamchik commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,142 @@
+    maxInstructionsCount = max(len(actualInstructions), len(expectedInstructions))
+
+    differences = []
+    for i in range(maxInstructionsCount):
+        actual = actualInstructions[i] if i < len(actualInstructions) else None
+        expected = expectedInstructions[i] if i < len(expectedInstructions) else None
+        if not actual and not expected:
+            raise ValueError("Invalid instructions data")
+        elif (actual is not None) and (
+            actual.get("bytecode") != expected.get("bytecode")
+        ):
+            differences.append(i)
```
The same is true for `actual is None` and some `expected` should also be added to the differences.

TheDZhon replied:
Might be solve with the following chunk (warn: untested)

```python
    if actual != expected:
        if actual is None or expected is None:
            differences.append(i)
        elif actual.get("bytecode") != expected.get("bytecode"):
            differences.append(i)
```

tamtamchik commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,142 @@
+    instructions = []
+    i = 0
+    while i < len(buffer):
+        opcode = buffer[i]
```
Consider adding an edge case for not valid opcode (e.g., `if opcode not in OPCODES`)

TheDZhon replied:
I'd say it should be a warning only

Alexvozhak commented on config_samples/lido_dao_holesky_config.json:
```diff
@@ -111,5 +112,41 @@
         }
+    },
+    "binary_checking": {
+	"enable": true,
```
done

Alexvozhak commented on config_samples/lido_dao_holesky_config.json:
```diff
@@ -111,5 +112,41 @@
+    },
+    "binary_checking": {
+	"enable": true,
+	"skip_deploy_error" : false,
```
if we want to find all binary comparison errors for all contracts in one pass

Alexvozhak commented on config_samples/lido_dao_holesky_config.json:
```diff
@@ -111,5 +112,41 @@
+    "binary_checking": {
+	"enable": true,
+	"skip_deploy_error" : false,
+	"hardhat_config_name": "holesky_hardhat.config.js",
```
do you mean the [same](https://github.com/lidofinance/diffyscan/pull/52#discussion_r1705345906) point?

Alexvozhak commented on diffyscan/diffyscan.py:
```diff
@@ -3,39 +3,121 @@
+        return
+
+    local_RPC_URL = config["binary_checking"]["local_RPC_URL"]
+    local_contract_address, text_reason = deploy_contract(
```
done

Alexvozhak commented on diffyscan/diffyscan.py:
```diff
@@ -149,28 +230,90 @@ def process_config(path: str, recursive_parsing: bool, unify_formatting: bool):
+        explorer_token = os.getenv("ETHERSCAN_EXPLORER_TOKEN", default=None)
+    if explorer_token is None:
+        raise ValueError(
+            f'Failed to find explorer token in env ("ETHERSCAN_EXPLORER_TOKEN")'
```
done

Alexvozhak commented on diffyscan/diffyscan.py:
```diff
@@ -149,28 +230,90 @@ def process_config(path: str, recursive_parsing: bool, unify_formatting: bool):
 
 
 def parse_arguments():
     parser = argparse.ArgumentParser()
```
done

Alexvozhak commented on diffyscan/diffyscan.py:
```diff
@@ -3,39 +3,121 @@
-from .utils.explorer import get_contract_from_explorer
-from .utils.github import get_file_from_github, get_file_from_github_recursive, resolve_dep
+from .utils.common import load_config, load_env, prettify_solidity
+from .utils.constants import *
```
done

Alexvozhak commented:
@TheDZhon fixed all remarks except encoder.py and binary_verifier.py. I will fix it for a few days

Alexvozhak commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,142 @@
+
+    actualInstructions = parse(actualBytecode)
+    expectedInstructions = parse(expectedBytecode)
+    maxInstructionsCount = max(len(actualInstructions), len(expectedInstructions))
```
done

Alexvozhak commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,142 @@
+
+    for ind in list(checkpoints):
+        startIndex = max(0, ind - nearLinesCount)
+        lastIndex = min(ind + nearLinesCount, maxInstructionsCount - 1)
```
done

Alexvozhak commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,142 @@
+        startIndex = max(0, ind - nearLinesCount)
+        lastIndex = min(ind + nearLinesCount, maxInstructionsCount - 1)
+        for i in range(startIndex, lastIndex + 1):
+            checkpoints.add(i)
```
done

TheDZhon commented on README.md:
````diff
@@ -35,6 +41,17 @@ Set your Github token to query API without strict rate limiting,
+export REMOTE_RPC_URL =<remote-rpc-url>
+```
+
+Set local RPC URL to validate contract bytecode at local rpc node (Hardhat),
````
This description is misleading, we need to have LOCAL_RPC_URL mostly to address the case when immutables should be checked against the local deployment and provided ctor arguments, isn't it?

Alexvozhak replied:
Set local RPC URL to check immutables against the local deployment and provided constructor arguments

TheDZhon commented on README.md:
```diff
@@ -62,9 +79,54 @@ Alternatively, create a new config file named `config.json`,
             "relative_root": "contracts"
         }
+    },
+    "raise_exception": true,
```
What does 'raise_exception' mean in this context, appearing out of the blue?

Alexvozhak replied:
renamed to raise_exceptions_on_comparison

TheDZhon replied:
1) It's still not clear — maybe it's better to describe the outcome, not the mechanism?
e.g.: `fail_on_comparison_error`
2) somewhere it's renamed, somewhere (like here) it's not :thinking:

TheDZhon commented on README.md:
```diff
@@ -62,9 +79,54 @@ Alternatively, create a new config file named `config.json`,
+    "bytecode_comparison": {
+        "hardhat_config_name": "holesky_hardhat.config.js",
+        "constructor_calldata": {
+            "0x28FAB2059C713A7F9D8c86Db49f9bb0e96Af1ef8": "000000000000000000000000ab89ed3d8f31bcf8bb7de53f02084d1e6f043d34000000000000000000000000e92329ec7ddb11d25e25b3c21eebf11f15eb325d00000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000"
+        },
+        "constructor_args": {
+            "0x28FAB2059C713A7F9D8c86Db49f9bb0e96Af1ef8": [
+                "0xab89ED3D8f31bcF8BB7De53F02084d1e6F043D34",
+                "0xE92329EC7ddB11D25e25b3c21eeBf11f15eB325d",
+                ""
```
Is it a good example when we have both args and calldata for the same contract address? If so, what would be the order of preference?

In my view, the program should treat this config being ambiguous and invalid (malformed) :warning:

Alexvozhak replied:
i've prefered prepared calldata from field 'constructor_calldata' to raw calldata in 'constructor_args'. Fixed example

TheDZhon replied:
not sure if it's a good approach still: it's too ambiguous and undocumented, why we can't enforce a uniqueness?

TheDZhon commented on README.md:
````diff
@@ -62,9 +79,54 @@ Alternatively, create a new config file named `config.json`,
     }
 }
 ```
+use `holesky_hardhat.config.js` from `config.json` or create near the diffyscan.py own Hardhat config `hardhat_config.js`: 
````
Suppose I installed diffyscan using the command from the Install section of README:
```
pipx install git+https://github.com/lidofinance/diffyscan
```

Where should I place this, then?

Alexvozhak replied:
the readme has been updated in this part

TheDZhon commented on config_samples/lido_dao_holesky_config.json:
```diff
@@ -1,115 +1,175 @@
+      "relative_root": "contracts"
+    }
+  },
+  "raise_exception": true,
```
I find this name being too vague tbh (as said above)

Alexvozhak replied:
raise_exceptions_on_comparison

TheDZhon commented on config_samples/lido_dao_holesky_config.json:
```diff
@@ -1,115 +1,175 @@
+  "raise_exception": true,
+  "bytecode_comparison": {
+    "hardhat_config_name": "holesky_hardhat_config.js",
+    "constructor_calldata": {
+      "0x28FAB2059C713A7F9D8c86Db49f9bb0e96Af1ef8": "000000000000000000000000ab89ed3d8f31bcf8bb7de53f02084d1e6f043d34000000000000000000000000e92329ec7ddb11d25e25b3c21eebf11f15eb325d00000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000"
+    },
+    "constructor_args": {
+      "0x28FAB2059C713A7F9D8c86Db49f9bb0e96Af1ef8": [
+        "0xab89ED3D8f31bcF8BB7De53F02084d1e6F043D34",
+        "0xE92329EC7ddB11D25e25b3c21eeBf11f15eB325d",
+        ""
+      ],
```
once again, it's a duplicate address

Alexvozhak replied:
the duplicate from 'constructor_args' has been removed

TheDZhon commented on config_samples/lido_dao_sepolia_config.json:
```diff
@@ -1,101 +1,298 @@
+  "bytecode_comparison": {
+    "hardhat_config_name": "sepolia_hardhat_config.js",
+    "constructor_calldata": {
+      "0x8f6254332f69557A72b0DA2D5F0Bc07d4CA991E7": "000000000000000000000000f30a674935479cc6f2254ba65d7534eab8bd6ba20000000000000000000000006885e36bfcb68cb383dfe90023a462c03bcb2ae500000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000"
+    },
+    "constructor_args": {
+      "0x8f6254332f69557A72b0DA2D5F0Bc07d4CA991E7": [
+        "0xf30a674935479cc6f2254ba65d7534eab8bd6ba2",
+        "0x6885E36BFcb68CB383DfE90023a462C03BCB2AE5",
+        ""
```
duplicate :thinking:

Alexvozhak replied:
fixed

TheDZhon commented on diffyscan/diffyscan.py:
```diff
@@ -3,39 +3,106 @@
+    if is_fully_matched:
+        return
+
+    logger.info(
+        f"Trying to check bytecode via using calldata and deploying into local node"
+    )
```
Shall say here that automated match hasn't worked out

Alexvozhak replied:
done

TheDZhon commented on diffyscan/diffyscan.py:
```diff
@@ -145,32 +211,127 @@ def run_diff(config, name, address, explorer_api_token, github_api_token, recurs
+    local_rpc_url = os.getenv("LOCAL_RPC_URL", "")
+    if not local_rpc_url:
+        raise ValueError("LOCAL_RPC_URL variable is not set")
+    if not local_rpc_url.startswith("http://"):
+        local_rpc_url = "http://" + local_rpc_url
```
why do we do this :) It's a bit surprising and inconsistent when looking down at the `REPOTE_RPC_URL` case, e.g.

Alexvozhak replied:
fixed

TheDZhon commented on diffyscan/utils/common.py:
```diff
@@ -57,3 +79,27 @@ def parse_repo_link(repo_link):
+
+def prettify_solidity(solidity_contract_content: str):
+    github_file_name = os.path.join(
+        tempfile.gettempdir(), "9B91E897-EA51-4FCC-8DAF-FCFF135A6963.sol"
```
Can't we generate a random uuid here for the filename?

Alexvozhak replied:
done

TheDZhon commented on diffyscan/utils/encoder.py:
```diff
@@ -0,0 +1,121 @@
+    if not bytes_str:
+        return to_hex_with_alignment(0)
+
+    # untested, taken from here:
```
huh?

Alexvozhak replied:
i'm waiting the example of specific contract

TheDZhon replied:
here: https://sepolia-optimism.etherscan.io/address/0xf49D208B5C7b10415C7BeAFe9e656F2DF9eDfe3B

init tx: https://sepolia-optimism.etherscan.io/tx/0x4c9db6f7c0e2ca4b831bfc67f46f265a1bdac5735738900e42014c88ef361373

TheDZhon replied:
https://docs.optimism.io/chain/networks#op-sepolia

TheDZhon commented on diffyscan/utils/encoder.py:
```diff
@@ -0,0 +1,121 @@
+
+
+def encode_constructor_arguments(constructor_abi: list, constructor_config_args: list):
+    # see https://docs.soliditylang.org/en/develop/abi-spec.html#contract-abi-specification
```
it's good to have this comment though might have been a bit more specific

Alexvozhak replied:
added [link ](https://github.com/lidofinance/lido-dao/blob/master/bytecode-verificator/bytecode_verificator.sh#L369-L405)

TheDZhon commented on README.md:
````diff
@@ -62,9 +80,54 @@ Alternatively, create a new config file named `config.json`,
     }
 }
 ```
+then create a new Hardhat config file named `hardhat_config.js` near the diffyscan.py
````
What if you installed diffyscan via pipx as suggested in README? Where should I place this file?

Alexvozhak replied:
done

TheDZhon commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,142 @@
+        return
+    logger.warn(f"Bytecodes have differences")
+
+    nearLinesCount = 3
```
doubling down

TheDZhon reviewed (changes requested):
:eyes:

TheDZhon commented on README.md:
```diff
@@ -62,9 +80,54 @@ Alternatively, create a new config file named `config.json`,
+    hardhat: {
+      chainId: 17000,
+      blockGasLimit: 92000000,
+      hardfork: "istanbul",
```
```suggestion
      hardfork: "cancun",
```

TheDZhon replied:
not critical though

Alexvozhak replied:
done

TheDZhon commented on package.json:
```diff
@@ -1,6 +1,16 @@
 {
```
maybe add `husky` for an automated setup of a precommit hook

Alexvozhak replied:
done

TheDZhon reviewed (commented):
:eyes:

Alexvozhak commented on diffyscan/utils/binary_verifier.py:
```diff
@@ -0,0 +1,159 @@
+    if unknown_opcodes:
+        logger.warn(f"Detected unknown opcodes: {unknown_opcodes}")
+
+    zipped_instructions = list(zip(actual_instructions, expected_instructions))
```
done

TheDZhon commented:
Seems the PR is ready to merge :star2: except only for two things:
1) An outstanding branch conflict it has
2) Metadata parsed as opcodes while it shouldn't, see https://docs.soliditylang.org/en/latest/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode
<img width="640" alt="image" src="https://github.com/user-attachments/assets/4ae5bb39-b53f-477c-bd06-37283a110458"> 

see as a reference:

> https://playground.sourcify.dev/
> address: 0x76A50b8c7349cCDDb7578c6627e79b5d99D24138
> network: #10 OP Mainnet
> <img width="640" alt="image" src="https://github.com/user-attachments/assets/41b4e248-f2ac-4252-9d23-a23d432cf91e">
>

TheDZhon commented:
https://github.com/lidofinance/aave-delivery-infrastructure/blob/main/verifications/scripts/verify-bytecode.ts#L479-L494
here is the example of the metadata parsing btw

TheDZhon reviewed (approved):
LGTM, accepting to upstream :green_apple:
